docs/reference/configuration/configuration.md

```diff
@@ -1,48 +1,276 @@
+
+Above, the configuration parameters `param1` and `param2` are defined using a description object. The description object supports the following keys:
+
+  - `help`: an optional help message that describes the purpose of the parameter.
+  - `value`: an optional field that defines the value of the parameter.
+  - `required`: an optional key that specifies whether the parameter must have a value before compiling the code (`false` by default). It's not possible to compile a source tree with one or more required parameters that don't have a value. Generally, setting `required` to true is only useful when `value` is not set.
+  - `macro_name`: an optional name for the macro defined at compile time for this configuration parameter. The configuration system automatically figures out the corresponding macro name for a configuration parameter, but the user can override this automatically computed name by specifying `macro_name`.
+
+You define a macro by value by using an integer or string instead of the description object, such as `param3` above. Defining a parameter by value is equivalent to defining it with a description object with the key `value` set to the value in place of the description object, the key `help` not set, the key `macro_name` not set and the key `required` set to `false`. Defining a parameter by value is equivalent to a configuration parameter defined with a description object with the key `value` set to the value in place of the description object, the key `help` unset, the key `macro_name` unset, and the key `required` set to `false`.
+
+<span class="notes">**Note:** The name of a parameter in `config` can't contain a dot (`.`) character.</span>
+
+The configuration system appends a prefix to the name of each parameter, so a parameter with the same name in a library does not conflict with parameters of the same name in targets or other libraries. The prefix is:
+
+| Location | Prefix |
+| -------- | ------ |
+| Target | `target.` |
+| Any library | The name of the library, as found in the `name` section of `mbed_lib.json`, followed by a dot (.) |
+| Application | `app.` |
 
-The default stack size of a user spawned thread is 4 kilobytes. You can configure the stack size of user spawned threads by using the [Mbed config system](/docs/development/tools/index.html#the-configuration-system) or passing in the [Thread constructor](/docs/development/reference/thread.html). You can reduce the default stack size for user spawned threads from 4K to 2K in the following two ways:
+### `mbed_lib.json` format specification
 
-1. Create an `mbed_app.json` file for your project with the following content:
+`mbed_lib.json` is a JSON formatted document that contains a root JSON Object. The keys within this object are sections. See the allowed sections and their meanings below:
```

Comment:
Not required simplification: Can we not call the keys in the object "sections"? Seems like an unnecessary new "concept" to introduce. Could the above:

> The keys within this object are sections. See the allowed sections and their meanings below:

be reworded to:

> The keys within this object and their meanings are found below:

And then change the "Section" header of the table below to "Key".

I notice you use "section" throughout this doc, but I think those uses are fine as it is informally referring to the "section" of the JSON document. I might be overanalyzing this a bit, but I thought I had missed a key config concept when I read the whole "section" bit.

Reply:
I'm on the fence about this, but maybe it would be good to maintain the concept. I understand what a "section" is in general, and now I'm being told that the topmost JSON keys are used to divide the root JSON object into sections.

Reply:
Plus then you have a word to distinguish between your topmost keys, and those lower in the hierarchy. For better or worse.

Reply:
A fair point! Definitely required, mostly just my knee-jerk reaction.